File: puzzles/medium/js/the-gift.js
let budgets = [];
let totalBudget = 0;

const NB_PARTICIPANTS = parseInt(readline());
let giftPrice = parseInt(readline());
for (let i = 0; i < NB_PARTICIPANTS; i++) {
  let budget = parseInt(readline());
  budgets.push(budget);
  totalBudget += budget;
}

if (totalBudget < giftPrice) {
  print("IMPOSSIBLE");
} else {
  budgets.sort(function(x, y) {
    return x - y;
  });
  for (let i = 0; i < NB_PARTICIPANTS; i++) {
    let contribution = budgets[i];
    let maxContribution = giftPrice / (NB_PARTICIPANTS - i);
    if (contribution >= maxContribution)
      contribution = Math.floor(maxContribution);
    print(contribution);
    giftPrice -= contribution;
  }
}

File: puzzles/medium/go/the-gift.go
package main

import (
  "fmt"
  "sort"
)
//import "os"

func main() {
  var budgets []int
  var totalBudget = 0

  var nbParticipants int
  fmt.Scan(&nbParticipants)
  
  var giftPrice int
  fmt.Scan(&giftPrice)
  
  for i := 0; i < nbParticipants; i++ {
    var budget int
    fmt.Scan(&budget)
    budgets = append(budgets, budget)
    totalBudget += budget
  }

  if totalBudget < giftPrice {
    fmt.Println("IMPOSSIBLE")
  } else {
    sort.Ints(budgets)
    for i := 0; i < nbParticipants; i++ {
      var contribution = budgets[i]
      var maxContribution = giftPrice / (nbParticipants - i)
      if contribution >= maxContribution {
        contribution = maxContribution
      }
      fmt.Println(contribution)
      giftPrice -= contribution
    }
  }
}
